### PR TITLE
util/rowcodec: add new row codec lib

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
Prepare to implement the new row format described in the [ proposal](https://github.com/pingcap/tidb/blob/master/docs/design/2018-07-19-row-format.md)

### What is changed and how it works?
Add the rowcodec library and tests, not used in TiDB yet.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7597)

### CODE REVIEW

**Comment by breeswish** [comment]
How is the progress of this PR?

**Comment by zz-jason** [comment]
@coocood friendly ping, any update?

**Comment by coocood** [comment]
@zz-jason 
I will refine this PR this week.

File: util/rowcodec/common.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (zz-jason) [inline util/rowcodec/common.go]:
```suggestion
// Copyright 2019 PingCAP, Inc.
```

File: util/rowcodec/common.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+// CodecVer is the constant number that represent the new row format.
+const CodecVer = 128
+
+var errInvalidCodecVer = errors.New("invalid codec version")
+
+// row is the struct type used to access the a row.
+type row struct {
+	// small:  colID []byte, offsets []uint16, optimized for most cases.
+	// large:  colID []uint32, offsets []uint32.
```
Comment (zz-jason) [inline util/rowcodec/common.go]: These two comment lines are better to be put on the comment of `row` struct.

File: util/rowcodec/decoder.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (zz-jason) [inline util/rowcodec/decoder.go]:
```suggestion
// Copyright 2019 PingCAP, Inc.
```

**Comment by coocood** [comment]
@zz-jason 
It's expensive to split into multiple PRs because basically there just two functions encode and decode, and we need both of them to run the test.

File: util/rowcodec/decoder.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	requestTypes  []*types.FieldType
+	origDefaults  [][]byte
+	loc           *time.Location
+}
+
+// NewDecoder creates a NewDecoder.
+// requestColIDs is the columnIDs to decode. tps is the field types for request columns.
+// origDefault is the original default value in old format, if the column ID is not found in the row,
+// the origDefault will be used.
+func NewDecoder(requestColIDs []int64, handleColID int64, tps []*types.FieldType, origDefaults [][]byte,
```
Comment (qw4990) [inline util/rowcodec/decoder.go]: Changing the type of `origDefaults` to `[]Datum` is more convenient for users?

Comment (coocood) [reply util/rowcodec/decoder.go]: The `origDefault` value is stored in the schema, which has type `[]byte`.

**Comment by jackysp** [comment]
Should we add some benchmark results in the description?

File: util/rowcodec/encoder.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+			r.offsets[i] = uint16(len(r.data))
+		}
+	}
+	buf = append(buf, CodecVer)
+	flag := byte(0)
+	if r.large {
+		flag = 1
+	}
+	buf = append(buf, flag)
+	buf = append(buf, byte(r.numNotNullCols), byte(r.numNotNullCols>>8))
```
Comment (qw4990) [inline util/rowcodec/encoder.go]: Using `binary.LittleEndian.PutUint16` is more clear and safe, which is consistent with `binary.LittleEndian.Uint16` in `setRowData`?

Comment (coocood) [reply util/rowcodec/encoder.go]: We use `append` here, using `binary.LittleEndian.PutUint16` require append two bytes first.

File: util/rowcodec/encoder.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+func (encoder *Encoder) build(buf []byte) ([]byte, error) {
+	r := &encoder.row
+	// Separate null and not-null column IDs.
+	numCols := len(encoder.tempColIDs)
+	nullIdx := numCols - int(r.numNullCols)
+	notNullIdx := 0
+	if r.large {
+		encoder.initColIDs32()
```
Comment (qw4990) [inline util/rowcodec/encoder.go]: No initialization for `r.data`?

Comment (coocood) [reply util/rowcodec/encoder.go]: We don't know how large the r.data would be, and the `r.data` will be reused, so we don't need to worry about the slice grow.

Comment (qw4990) [reply util/rowcodec/encoder.go]: My fault, I forget it has been reset at the begging of `Encode`.

File: util/rowcodec/common.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+const CodecVer = 128
+
+var errInvalidCodecVer = errors.New("invalid codec version")
+
+// row is the struct type used to access the a row.
+// There are two types of row, small and large.
+// small rows use one byte colID and two bytes offset, optimized for most cases.
+// large rows use four bytes colID and four bytes offset.
+type row struct {
+	large          bool
```
Comment (zz-jason) [inline util/rowcodec/common.go]: s/large/isLarge/?

File: util/rowcodec/common.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/pingcap/errors"
+)
+
+// CodecVer is the constant number that represent the new row format.
+const CodecVer = 128
+
+var errInvalidCodecVer = errors.New("invalid codec version")
+
+// row is the struct type used to access the a row.
+// There are two types of row, small and large.
```
Comment (zz-jason) [inline util/rowcodec/common.go]: It's better to explain more on what kind of row is considered as small or large row.

File: util/rowcodec/common.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+	// for large row
+	colIDs32  []uint32
+	offsets32 []uint32
+}
+
+func (r *row) getData(i int) []byte {
+	var start, end uint32
+	if r.large {
+		if i > 0 {
```
Comment (zz-jason) [inline util/rowcodec/common.go]: can we using one more bucket for `offsets`? like Chunk.

File: util/rowcodec/common.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+		cursor += colIDsLen
+		offsetsLen := int(r.numNotNullCols) * 2
+		r.offsets = bytes2U16Slice(rowData[cursor : cursor+offsetsLen])
+		cursor += offsetsLen
+	}
+	r.data = rowData[cursor:]
+	return nil
+}
+
+func bytesToU32Slice(b []byte) []uint32 {
```
Comment (zz-jason) [inline util/rowcodec/common.go]: I remembered that we have a lot of places use these kind of functions, can we put them together in the `util` package?

File: util/rowcodec/encoder.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+	encoder.values = append(encoder.values, d)
+}
+
+// Encode encodes a row from a datums slice.
+func (encoder *Encoder) Encode(colIDs []int64, values []types.Datum, buf []byte) ([]byte, error) {
+	encoder.reset()
+	for i, colID := range colIDs {
+		encoder.addColumn(colID, values[i])
+	}
+	return encoder.build(buf[:0])
```
Comment (zz-jason) [inline util/rowcodec/encoder.go]: It's better to let the `Encode()` caller to reset the `buf` when it pass the parameter to this function.

File: util/rowcodec/common.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	var b []byte
+	hdr := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+	hdr.Len = len(u32s) * 4
+	hdr.Cap = hdr.Len
+	hdr.Data = uintptr(unsafe.Pointer(&u32s[0]))
+	return b
+}
+
+func encodeInt(buf []byte, iVal int64) []byte {
```
Comment (zz-jason) [inline util/rowcodec/common.go]: can we use `codec. EncodeInt()`?

File: util/rowcodec/common.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+import (
+	"encoding/binary"
+	"reflect"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+)
+
+// CodecVer is the constant number that represent the new row format.
+const CodecVer = 128
```
[the remark this answers is not in the document]
Comment (coocood) [reply util/rowcodec/common.go]: It doesn't have to be 128.
The number must not conflict the first byte of the old row format.
Currently, the largest flag number is 10, we need to reserve some number for future types.

File: util/rowcodec/common.go (hunk 9 above)
Comment (coocood) [reply util/rowcodec/common.go]: We use `bytesToU32Slice` to convert bytes in the encoded row to `[]uint32` slice, so an extra 0 element would increase the row size, wastes storage space.

File: util/rowcodec/common.go (hunk 10 above)
Comment (coocood) [reply util/rowcodec/common.go]: Maybe not used in the same way?
We can do it in future PR.

File: util/rowcodec/common.go (hunk 12 above)
Comment (coocood) [reply util/rowcodec/common.go]: `codec.EncodeInt` always encodes to 8 bytes.
This `encodeInt` can encode to 1, 2, 4, 8 bytes.

**Comment by coocood** [comment]
@zz-jason PTAL

File: util/rowcodec/encoder.go [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+	encoder.tempColIDs = append(encoder.tempColIDs, colID)
+	encoder.values = append(encoder.values, d)
+}
+
+// Encode encodes a row from a datums slice.
+func (encoder *Encoder) Encode(colIDs []int64, values []types.Datum, buf []byte) ([]byte, error) {
+	encoder.reset()
+	for i, colID := range colIDs {
+		encoder.addColumn(colID, values[i])
+	}
```
Comment (zz-jason) [inline util/rowcodec/encoder.go]: After this for loop:
1. `encoder.tempColIDs` is a copy of `colIDs`.
2. `encoder.values` is a copy of `values`.
3. `encoder.numNullCols` and `encoder.numNotNullCols` is calculated based on `values`.

The first two result can be obtain by directly calling a `append()` function, it's much efficient than the original way I think.

File: util/rowcodec/encoder.go [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+			} else {
+				r.colIDs[notNullIdx] = byte(colID)
+			}
+			encoder.values[notNullIdx] = encoder.values[i]
+			notNullIdx++
+		}
+	}
+	if r.isLarge {
+		largeNotNullSorter := (*largeNotNullSorter)(encoder)
+		sort.Sort(largeNotNullSorter)
```
Comment (zz-jason) [inline util/rowcodec/encoder.go]: why do we need to sort `colIDs` and the `values` for each row?

Comment (coocood) [reply util/rowcodec/encoder.go]: So we can do binary searches later.

**Pull request merged**